tests/group/group_constructors.cpp

```diff
@@ -91,53 +142,114 @@ class TEST_NAME : public util::test_base {
               simpleRange, simpleRange, [=](cl::sycl::group<numDims> group) {
-                // Check copy constructor
-                cl::sycl::group<numDims> copied(group);
-                check_equality(successAcc, current_check::copy_constructor,
-                               copied, group);
-
-                // Check move constructor
-                cl::sycl::group<numDims> moved(std::move(copied));
-                check_equality(successAcc, current_check::move_constructor,
-                               moved, group);
+                const auto& groupReadOnly = group;
+                state_storage<numDims> expected(groupReadOnly);
 
+                // Check copy constructor
+                {
+                  cl::sycl::group<numDims> copied(groupReadOnly);
+                  check_equality(successAcc, current_check::copy_constructor,
+                                 copied, expected);
+                }
                 // Check copy assignment
-                copied = moved;
-                check_equality(successAcc, current_check::copy_assignment,
-                               copied, group);
+                {
+                  auto copied = groupReadOnly;
+                  check_equality(successAcc, current_check::copy_assignment,
+                                 copied, expected);
+                }
+                // Check move constructor; invalidates group
+                {
+                  cl::sycl::group<numDims> moved(std::move(group));
+                  check_equality(successAcc, current_check::move_constructor,
+                                 moved, expected);
+                }
+              });
+        });
+        testQueue.submit([&](cl::sycl::handler& cgh) {
+          auto successAcc =
+              successBuf.get_access<cl::sycl::access::mode::write>(cgh);
 
-                // Check move assignment
-                moved = std::move(copied);
+          cgh.parallel_for_work_group<group_move_assignment_kernel<numDims>>(
+              simpleRange, simpleRange, [=](cl::sycl::group<numDims> group) {
+                state_storage<numDims> expected(group);
+
+                // Check move assignment; invalidates group
+                auto moved = std::move(group);
                 check_equality(successAcc, current_check::move_assignment,
-                               moved, group);
+                               moved, expected);
               });
         });
+        testQueue.wait_and_throw();
+      }
+
+      // Check on the host side only if copy assignment works as expected
+      if (success[to_integral(current_check::copy_assignment)]) {
+        // group is not default constructible, store two objects into the array
+        static constexpr size_t numItems = 2;
+        using item_array_t = std::array<cl::sycl::group<numDims>, numItems>;
+        char rawItems[sizeof(item_array_t)];
+        auto& items = *reinterpret_cast<item_array_t*>(rawItems);
```

Comment:
OK, now I see the hack to create a `std::array` without a default constructor... :-)

I am worried about alignment constraints on `item_array_t` here that might not be the same as a byte in all the ABIs.

Reply:
Do we say in the spec that these objects are trivially copyable, by the way?

So I am not sure we can use accessors of `group` for example...

Reply:
> Each of the following SYCL runtime classes: id, range, item, nd_item, h_item, group, sub_group and nd_range
> must follow the following statements, where T is the runtime class type:
> • T must be default copy constructible and copy assignable on the host application (in the case where T is
> available on the host) and within SYCL kernel functions.
> • T must be default destructible on the host application (in the case where T is available on the host) and
> within SYCL kernel functions.
> • T must be default move constructible and default move assignable on the host application (in the case where
> T is available on the host) and within SYCL kernel functions.
> • T must be equality comparable on the host application (in the case where T is available on the host) and
> within SYCL kernel functions. Equality between two instances of T (i.e. a == b) must be true if the value
> of all members are equal and non-equality between two instances of T (i.e. a != b) must be true if the
> value of any members are not equal, unless either instance has become invalidated by a move operation. By
> extension of the requirements above, equality on T must guarantee to be reflexive (i.e. a == a), symmetric
> (i.e. a == b implies b == a and a != b implies b != a) and transitive (i.e. a == b && b == c implies
> c == a).

This seems to imply "trivially copiable" requirement.

NOTE: existing code already uses "the hack", so no new requirements are enforced. @keryell, I suggest filing an separate issue, if this question requires additional clarification.

Reply:
Great catch regarding alignment. I've fixed it by 5ef2c768737b0cc7236c46f6c0984097fc47b5ca.
I've also moved all the storage logic to the same place by 5a98c20c87bd68dca67b7d15faa8237656ca6391, so may be it will be less confusing to read.